source/solutions-library/fin-services-fireworks-rag.txt

```diff
@@ -0,0 +1,550 @@
+
+      .. code-block:: python
+         :copyable: true
+
+         firectl deploy ccmodel
+         
+The seamless integration of MongoDB and Fireworks.AI streamlines the
+process of enhancing AI model performance in a cost-effective and
+efficient manner. This is evident in the step-by-step guide to
+fine-tuning provided above. 
+
+After deploying the model in the fireworks platform as a serverless API,
+the model id (models/ft-m88hxaga-pi11m) shown in Fig 2 can be used to
+invoke the fine-tuned SLM model using the language model framework of
+your choice.
+
+
+Key Learnings
+-------------
+
+Achievements and Potential Outcomes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Fine-tuning the SLM for credit card recommendations has yielded
+impressive results:
+
+Applying the above fine-tuning steps to the SLM model in the context of
+of the alternative credit card product recommendation process, we
+observed a tremendous improvement in the response time of the process
+using the fine-tuned SLM:
+
+**1. Response Time Improvement**: Originally, the response time with a
+large LLM would be around 5 seconds. By using an SLM, this can be
+reduced to approximately 19x reduction in latency.
+
+.. code-block:: python
+   :copyable: true
+
+    import time
+    class tiktok:
+    """
+    Decorator to time the execution of a function and log the time taken.
+    """
+    def __init__(self, function):
+          self.function = function
+
+    def __call__(self, *args, **kwargs):
+          import time
+          start = time.time()
+          value = self.function(*args, **kwargs)
+          end = time.time()
+          print(f"Time taken for {self.function.__name__}: {end - start} seconds")
+          return value
+
+    @tiktok
+    @mdbcache
+    def invoke_llm(prompt):
+    """
+    Invoke the language model with the given prompt with cache. The invole LLM method can invoke either a LLM or SLM based on the Fireworks Model ID initialised.
+
+    Args:
+          prompt (str): The prompt to pass to the LLM.
+    """
+    ...
+
+
+.. list-table:: Model Inference Times
+   :header-rows: 1
+   :widths: 28 18 18 18 18
+
+   * - Model
+     - Inference Time 1 (s)
+     - Inference Time 2 (s)
+     - Inference Time 3 (s)
+     - Average Time (s)
+
+   * - llama-v3p1-405b-instruct
+     - 5.5954
+     - 7.5936
+     - 4.9121
+     - 6.0337
+
+   * - SLM - fine-tuned llama-v3p1-8b
+     - 0.3554
+     - 0.0480
+     - 0.0473
+     - 0.1502
+
+**2. Memory Reduction**:  LLMs typically require substantial memory, often in the range of  8x80GB VRAM. A SLM can operate efficiently with around 16 GB VRAM of memory, leading to a 97.5% reduction in memory usage.
+
+**3. Hardware Reduction**: Deploying LLMs often necessitates high-end GPUs or multiple servers. With an SLM, the hardware requirements can be significantly reduced, potentially allowing deployment on standard CPUs or a single server, leading to a drastic reduction in hardware costs.
+
+Conclusion
+~~~~~~~~~~
```

Comment:
(suggestion)
Would you like to uplevel this so it's at the same level as Key Learnings?
Conclusion 
-----------

Reply:
Thanks @andreaalaman! You're totally correct. I paged through some of the other solutions and we're not super consistent yet. We can adjust as we're doing a copy edit pass a little later on. I'll leave this one to your preference.